src/keys/age.rs

```diff
@@ -0,0 +1,799 @@
+}
+
+/// Encode header and encrypt payload given all the secrets and random inputs.
+/// The length of the output can be be computed with `enc_len`.
+///
+/// The crucial security parameter (besides password strength) is `work_factor`.
+/// Too small work factor (<15) will result in weak key derivation.
+/// Too large work factor (>25) will take too long to derive key.
+/// Recommended minimal value is `RECOMMENDED_MINIMUM_ENCRYPT_WORK_FACTOR`.
+/// `work_factor` must be <64.
+pub fn enc(
+    password: &[u8],
+    salt: &[u8; 16],
+    file_key: &[u8; 16],
+    work_factor: u8,
+    nonce: &[u8; 16],
+    plaintext: &[u8],
+    age: &mut [u8],
+) {
+    assert_eq!(enc_len(work_factor, plaintext.len()), age.len());
+    let h = header_len(work_factor);
+    enc_header(password, salt, file_key, work_factor, &mut age[..h]);
+    enc_payload(file_key, nonce, plaintext, &mut age[h..]);
+}
+
+/// Encode header and encrypt payload given all the secrets and random inputs producing a vector.
+#[cfg(feature = "std")]
+pub fn enc_vec(
+    password: &[u8],
+    salt: &[u8; 16],
+    file_key: &[u8; 16],
+    work_factor: u8,
+    nonce: &[u8; 16],
+    plaintext: &[u8],
+) -> Vec<u8> {
+    let mut age = Vec::new();
+    age.resize(enc_len(work_factor, plaintext.len()), 0_u8);
+    let h = enc_header(password, salt, file_key, work_factor, &mut age[..]);
+    enc_payload(file_key, nonce, plaintext, &mut age[h..]);
+    age
+}
+
+/// The recommended minimum work factor used by `encrypt`, or roughly 1 sec on modern PC (2023).
+pub const RECOMMENDED_MINIMUM_ENCRYPT_WORK_FACTOR: u8 = 19;
+
+/// Generate random salt, file key, and nonce and use them to protect plaintext in age format.
+///
+/// The crucial security parameter (besides password strength) is `work_factor`.
+/// Too small work factor (<15) will result in weak key derivation.
+/// Too large work factor (>25) will take too long to derive key.
+/// Recommended minimal value is `RECOMMENDED_MINIMUM_ENCRYPT_WORK_FACTOR`.
+/// `work_factor` must be <64.
+#[cfg(feature = "random")]
+pub fn encrypt(password: &[u8], work_factor: u8, plaintext: &[u8], age: &mut [u8]) {
+    let mut salt = [0_u8; 16];
+    let mut file_key = [0_u8; 16];
+    let mut nonce = [0_u8; 16];
+    crate::utils::rand::fill(&mut salt[..]).unwrap();
+    crate::utils::rand::fill(&mut file_key[..]).unwrap();
+    crate::utils::rand::fill(&mut nonce[..]).unwrap();
+    enc(password, &salt, &file_key, work_factor, &nonce, plaintext, age);
+    nonce.zeroize();
+    file_key.zeroize();
+    salt.zeroize();
+}
+
+/// Generate random salt, file key, and nonce and use them to protect plaintext in age format producing a vector.
+#[cfg(all(feature = "random", feature = "std"))]
+pub fn encrypt_vec(password: &[u8], plaintext: &[u8]) -> Vec<u8> {
+    let mut salt = [0_u8; 16];
+    let mut file_key = [0_u8; 16];
+    let mut nonce = [0_u8; 16];
+    crate::utils::rand::fill(&mut salt[..]).unwrap();
```

Comment:
Shall we return Result here?

Reply:
Wherever we can yes, people can recover from it, with a panic they can't